### DISCUSSION TITLE
In Simulation of tortuosity_fd, module 'openpnm.phase' has no attribute 'GenericPhase'. How to handle it?

### DISCUSSION THREAD

**Discussion by eromanga233** [PMEAL/porespy#741] [Q&A] [answered]
In Simulation of tortuosity_fd, I follows the instructions of the [Tutorials.](https://porespy.org/examples/simulations/tutorials/finding_tortuosity_from_image.html) However, while running the codes, there's always an error:
_AttributeError: module 'openpnm.phase' has no attribute 'GenericPhase'_
After checking the original version of openpnm-test-3.0.0, I find that there's indeed a module named 'GenericPhase' in it. But there's no module called 'GenericPhase' in current version of openpm, which may be the cause of the error mentioned. How can I use porespy to compute tortuosity now?

**Comment by ma-sadeghi** [comment, marked as answer]
There are two things:

- `openpnm-test` as the name suggests is not the actual package. Try installing `openpnm` (without the `-test`). 
- However, even when installing the actual OpenPNM v3, it's stil going to fail. I just pushed a fix and made a new minor release PoreSpy v2.2.1. Try updating your PoreSpy see if the problem is fixed now.

Sorry for the inconvenience.

**Reply by eromanga233** [reply]
- After upgrading the porespy, when I try to compute the tortuosity by following the instructions on porespy.org:
`results = ps.simulations.tortuosity_fd(im=img, axis=1)`
there's another error now:
`pypardiso.pardiso_wrapper.PyPardisoError: The Pardiso solver failed with error code -3. See Pardiso documentation for details.`
Please help me with this porblem.

**Reply by ma-sadeghi** [reply]
What's your MKL and pypardiso versions? Paste the output of "pip freeze" and "conda list" (you need to run in terminal/command line)

**Reply by eromanga233** [reply]
The version of MKL is 2022.2.0, and the version of pypardiso is 0.4.1.
I am still trying to find out what causes the error:
`pypardiso.pardiso_wrapper.PyPardisoError: The Pardiso solver failed with error code -3. See Pardiso documentation for details.`

**Reply by eromanga233** [reply]
Thankyou for your patience. After following the suggestion post on stack [overflow,](https://stackoverflow.com/questions/70665142/pypardisoerror-the-pardiso-solver-failed-with-error-code-3-see-pardiso-docum) I have solved the problem.
`pip install intel-openmp==2021.4`

**Reply by ma-sadeghi** [reply]
Great, glad you figured it out.
